**One-bin A <-> 3B reaction, with 1st-order kinetics in both directions,
taken to equilibrium**

Diffusion not applicable (just 1 bin)

LAST REVISED: Aug. 7, 2022

* [First Step](#sec_1)
* [Numerous more steps](#sec_2)
* [Equilibrium](#sec_3)

In [1]:
import set_path
set_path.add_ancestor_dir_to_syspath(3)  # The number of levels to go up 
                                         # to reach the project's home, from the folder containing this notebook

Added 'D:\Docs\- MY CODE\BioSimulations\life123-Win7' to sys.path


In [2]:
from experiments.get_notebook_info import get_notebook_basename

from modules.chemicals.chemicals import Chemicals as chem
from modules.reactions.reactions import Reactions
from life_1D.bio_sim_1d import BioSim1D as bio

from modules.html_log.html_log import HtmlLog as log
from modules.visualization.graphic_log import GraphicLog

In [3]:
# Initialize the HTML logging
log_file = get_notebook_basename() + ".log.htm"    # Use the notebook base filename for the log file

# Set up the use of some specified graphic (Vue) components
GraphicLog.config(filename=log_file,
                  components=["vue_cytoscape_1"],
                  extra_js="https://cdnjs.cloudflare.com/ajax/libs/cytoscape/3.21.2/cytoscape.umd.js",
                  home_rel_path="../../..")    # relative path is from the location of THE LOG FILE to the project's home

-> Output will be LOGGED into the file 'reaction_2.log.htm'


In [4]:
# Initialize the system
chem_data = chem(names=["A", "B"])     # NOTE: Diffusion not applicable (just 1 bin)

rxn = Reactions(chem_data)

# Reaction A -> 3B , with 1st-order kinetics in both directions
rxn.add_reaction(reactants=["A"], products=[(3,"B")], forward_rate=5., reverse_rate=2.)

bio.initialize_system(n_bins=1, chem_data=chem_data, reactions=rxn)

bio.set_uniform_concentration(species_index=0, conc=10.)
bio.set_uniform_concentration(species_index=1, conc=50.)

bio.describe_state()

1 bins and 2 species:
  Species 0 (A). Diff rate: NOT SET. Conc:  [10.]
  Species 1 (B). Diff rate: NOT SET. Conc:  [50.]


In [5]:
rxn.describe_reactions()

Number of reactions:  1
0: A <-> 3 B  (Rf = 5.0 / Rb = 2.0)


['0: A <-> 3 B  (Rf = 5.0 / Rb = 2.0)']

In [6]:
# Send the plot to the HTML log file
graph_data = rxn.prepare_graph_network()
GraphicLog.export_plot(graph_data, "vue_cytoscape_1")


 {'reactants': [(1, 0, 1)], 'products': [(3, 1, 1)], 'Rf': 5.0, 'Rb': 2.0} 

[Graphic element sent to log file]


### <a name="sec_1"></a>First step

In [7]:
# First step
bio.react(time_step=0.1, n_steps=1)
bio.describe_state()

1 bins and 2 species:
  Species 0 (A). Diff rate: NOT SET. Conc:  [15.]
  Species 1 (B). Diff rate: NOT SET. Conc:  [35.]


_Early in the reaction :_  
[A] = 15.   [B] = 35.

### <a name="sec_2"></a>Numerous more steps

In [8]:
# Numerous more steps
bio.react(time_step=0.1, n_steps=10)

bio.describe_state()

1 bins and 2 species:
  Species 0 (A). Diff rate: NOT SET. Conc:  [14.54545455]
  Species 1 (B). Diff rate: NOT SET. Conc:  [36.36363636]


### <a name="sec_3"></a>Equilibrium

Consistent with the 5/2 ratio of forward/reverse rates (and the 1st order reactions),
the systems settles in the equilibrium:   [A] = 14.54545455 , [B] = 36.36363636

In [9]:
A_eq = bio.bin_concentration(0, 0)
B_eq = bio.bin_concentration(0, 1)
print(f"Ratio of equilibrium concentrations: {B_eq / A_eq}")
print(f"Ratio of forward/reverse rates: {rxn.get_forward_rate(0) / rxn.get_reverse_rate(0)}")

Ratio of equilibrium concentrations: 2.4999999999828124
Ratio of forward/reverse rates: 2.5
